modules/lily-compiler/src/main/java/io/github/tomboyo/lily/compiler/icg/OasSchemaToAst.java

```diff
@@ -270,11 +262,95 @@ class package.MyClass,
         Stream.concat(Stream.of(exteriorClass), interiorAst));
   }
 
+  /* Gets properties from both properties keywords and composed schema */
+  @SuppressWarnings({"rawtypes", "unchecked"}) // not shit I can do about the io.swagger API
+  private Map<String, Schema> getProperties(Schema schema) {
+    var properties = new HashMap<String, Schema>();
+    if (schema.getProperties() != null) {
+      properties.putAll(schema.getProperties());
+    }
+
+    if (schema instanceof ComposedSchema c) {
+      Stream.of(c.getAllOf(), c.getAnyOf(), c.getOneOf())
+          .filter(Objects::nonNull)
+          .flatMap(List::stream)
+          .forEach(s -> properties.putAll(getProperties(s)));
+    }
+
+    return properties;
+  }
+
+  private Set<String> getMandatoryPropertyNames(Schema<?> root) {
+    var required = requiredPropertyNames(root);
+    var nonNullable = getNonNullablePropertyNames(root);
+    var mandatory = intersection(required, nonNullable);
+
+    if (root instanceof ComposedSchema c) {
+      /* If all the oneOf schema (one of which MUST validate) agree that a given property is mandatory, then it is
+      mandatory on the composed schema as well. */
+      Stream.ofNullable(c.getOneOf())
+          .flatMap(List::stream)
+          .map(this::getMandatoryPropertyNames)
+          .reduce(this::intersection)
+          .ifPresent(mandatory::addAll);
+
+      /* If a property is mandatory according to an allOf component, then it is mandatory according to the composed
+      schema as well. */
+      Stream.ofNullable(c.getAllOf())
+          .flatMap(List::stream)
+          .map(this::getMandatoryPropertyNames)
+          .forEach(mandatory::addAll);
+    }
+
+    return mandatory;
+  }
+
+  private Set<String> requiredPropertyNames(Schema<?> root) {
+    var required = new HashSet<>(requireNonNullElse(root.getRequired(), List.of()));
+
+    /* AllOf component required keywords are flattened into the composed schema. */
+    if (root instanceof ComposedSchema c) {
+      Stream.ofNullable(c.getAllOf())
+          .flatMap(List::stream)
+          .map(s -> (Schema<?>) s) // *internal screaming*
+          .map(this::requiredPropertyNames)
+          .flatMap(Set::stream)
+          .forEach(required::add);
+
+      Stream.ofNullable(c.getOneOf())
+          .flatMap(List::stream)
+          .map(s -> (Schema<?>) s) // *additional internal screaming*
+          .map(this::requiredPropertyNames)
+          .reduce(this::intersection)
+          .ifPresent(required::addAll);
+    }
+
+    return required;
+  }
+
+  private <T> HashSet<T> intersection(Collection<T> a, Collection<T> b) {
+    var result = new HashSet<>(a);
+    result.retainAll(b);
+    return result;
+  }
+
+  private Set<String> getNonNullablePropertyNames(Schema<?> root) {
```

Comment:
I think this should be recursive as well; if an allOf component says a property is non-nullable, then it is on the composed schema as well. I.e.
```
Foo:
  required: ['a']
  allOf:
    - properties
        a:
           type: string
```
Foo has a mandatory property 'a'. This can be the case when OneOf's have consensus on nullability as well.